### PR TITLE
spv_parsed_instruction_t cleanup

### PR DESCRIPTION
**Pull request by dneto0**
Add members:
 - words: a pointer to an array of words in the instruction,
   in host native endianness.
 - num_words: sizes the words member

Remove member:
 - offset

This simplifies clients of spvBinaryParse, because they don't
have to handle endianness translation.

Also, it makes the binary parse API more composable, allowing
for easy chaining of binary parse clients.  A binary parse client
is handed the array of words directly instead of having to reference
some external array of all the words in the SPIR-V binary.  It also
allows a binary parse client to mutate the instruction stream before
handing off to a downstream consumer.

TODO(dneto): Still need to write the unit tests for spvBinaryParse

Fixes: https://github.com/KhronosGroup/SPIRV-Tools/issues/1

### CODE REVIEW

File: source/binary.cpp [hunk 1]
```diff
@@ -136,13 +137,17 @@ class Parser {
...
 
-  // Parses an instruction operand with the given type.
-  // May update the expected_operands parameter, and the scalar members of the
-  // inst parameter. On success, returns SPV_SUCCESS, advances past the
-  // operand, and pushes a new entry on to the operands vector.  Otherwise
-  // returns an error code and issues a diagnostic.
-  spv_result_t parseOperand(spv_parsed_instruction_t* inst,
+  // Parses an instruction operand with the given type, for an instruction
+  // starting at inst_offset words into the SPIR-V binary.  Appends, to the
+  // words parameter, the endian-translated words for the operand, but only if
```
Comment (dmircevski) [inline source/binary.cpp]: Isn't there a simpler interface for this?  I'm finding it hard to follow.

File: source/binary.cpp [hunk 2]
```diff
@@ -281,13 +292,25 @@ spv_result_t Parser::parseInstruction() {
   // The zero values for all members except for opcode are the
   // correct initial values.
   spv_parsed_instruction_t inst = {};
-  inst.offset = _.word_index;
+  const size_t inst_offset = _.word_index;
```
Comment (antiagainst) [inline source/binary.cpp]: I'd prefer to move this variable to just before where it is used.

File: source/binary.cpp (hunk 1 above)
Comment (antiagainst) [inline source/binary.cpp]: This becomes a fairly complicated interface...

File: source/binary.cpp [hunk 3]
```diff
@@ -338,21 +363,33 @@ spv_result_t Parser::parseInstruction() {
...
   }
+  assert(inst_word_count == words.size());
 
-  recordNumberType(&inst);
+  recordNumberType(inst_offset, &inst);
 
-  // Must wait until here to set the inst.operands pointer because the vector
-  // might be resized while we accumulate itse elements.
+  if (_.requires_endian_conversion) {
+    // We must wait until here to set this pointer, because the vector might
```
Comment (antiagainst) [inline source/binary.cpp]: Nice explanation!

File: source/endian.h [hunk 4]
```diff
@@ -43,4 +43,7 @@ uint64_t spvFixDoubleWord(const uint32_t low, const uint32_t high,
 spv_result_t spvBinaryEndianness(const spv_const_binary binary,
                                  spv_endianness_t* endian);
 
+// Returns true if the given endianness the same as the host's native endiannes.
```
Comment (antiagainst) [inline source/endian.h]: ... if the given endianness *is* the same ...

File: source/binary.cpp (hunk 1 above)
Comment (dneto0) [inline source/binary.cpp]: The tipping point on the complexity of this is caused by the fact that the method behaves differently based on whether endian conversion must be done or not.  If no endian conversion is required, then the endian_converted_inst_words parameter is ignored.  Life would be simpler if we didn't mind copying the underlying data in that case.  We save on allocations and copying time in that very common case.  I think it's worth keeping that optimization.

Maybe I worded it poorly.  How about replacing the sentence from "Appends....from host" with:

If the SPIR-V binary is the same endianness as the host, then the endian_converted_inst_words parameter is ignored.  Otherwise, this method appends the words for this operand, converted to host native endianness, to the end of endian_converted_inst_words.

Comment (antiagainst) [inline source/binary.cpp]: +1. I like the later version.

File: source/binary.cpp (hunk 2 above)
Comment (dneto0) [inline source/binary.cpp]: _.word_index is modified later in this method and by calls to parseOperand.  So we need to save it early.
But yes, I can move it down.

File: source/endian.h (hunk 4 above)
Comment (dneto0) [inline source/endian.h]: will fix.

File: source/binary.cpp (hunk 1 above)
Comment (dneto0) [inline source/binary.cpp]: changed to new version

File: source/binary.cpp (hunk 2 above)
Comment (dneto0) [inline source/binary.cpp]: done

File: source/endian.h (hunk 4 above)
Comment (dneto0) [inline source/endian.h]: fixed, by saying "matches"

**Comment by antiagainst** [comment]
+2

**Comment by dneto0** [comment]
Rebased, squashed, and pushed onto master
